pkg/config/lifecycle_provider.go

```diff
@@ -0,0 +1,90 @@
+package config
+
+import (
+	"sync/atomic"
+
+	"github.com/google/go-containerregistry/pkg/authn"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	LifecycleConfigName = "lifecycle-image"
+	LifecycleConfigKey  = "image"
+)
+
+type RegistryClient interface {
+	Fetch(keychain authn.Keychain, repoName string) (v1.Image, string, error)
+}
+
+type lifecycleData struct {
+	image v1.Image
+	err   error
+}
+
+type LifecycleProvider struct {
+	RegistryClient RegistryClient
+	Keychain       authn.Keychain
+	lifecycleData  atomic.Value
+	callbacks      []func()
+}
+
+func NewLifecycleProvider(lifecycleImageRef string, client RegistryClient, keychain authn.Keychain) *LifecycleProvider {
+	p := &LifecycleProvider{
+		RegistryClient: client,
+		Keychain:       keychain,
+	}
+
+	data := &lifecycleData{}
+
+	data.image, data.err = p.fetchImage(lifecycleImageRef)
+
+	p.lifecycleData.Store(data)
+	return p
+}
+
+func (l *LifecycleProvider) UpdateImage(cm *corev1.ConfigMap) {
+	data := &lifecycleData{}
+	defer l.callCallBacks()
+	defer l.lifecycleData.Store(data)
+
+	imageRef, ok := cm.Data[LifecycleConfigKey]
+	if !ok {
+		data.err = errors.New("lifecycle-image config invalid")
+		return
+	}
+
+	data.image, data.err = l.fetchImage(imageRef)
+}
```

Comment:
when the image is invalid, are we not updating the value and doing callbacks? do we need to do it?

Reply:
I see from the test that this is intentional behaviour, but haven't figured out why we want to do a callback...